File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.cpp
#include <bits/stdc++.h>
using namespace std;

vector<int> solution(vector<string> genres, vector<int> plays) {
    vector<int> answer;
    map<string,vector<pair<int,int>>> m;
    map<string,int> cnt;
    for(int i=0;i<genres.size();i++){
        m[genres[i]].push_back({-plays[i],i});
        cnt[genres[i]]+=plays[i];
    }
    vector<pair<string,int>> tmp;
    for(auto x: cnt) tmp.push_back(x);
    sort(tmp.begin(),tmp.end(),[&](pair<string,int> a, pair<string,int> b){
         return a.second > b.second;
    });
    
    for(auto& [a,b]: m){
        sort(b.begin(),b.end());
    }
    for(auto x: tmp){
        for(int i=0;i<min((int)m[x.first].size(),2);i++){
            answer.push_back(m[x.first][i].second);
        }
    }

    return answer;
}

File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.java
import java.util.*;

class Solution {
    public List<Integer> solution(String[] genres, int[] plays) {
        Map<String, List<int[]>> genre = new HashMap<>();
        Map<String, Integer> cnt = new HashMap<>();
        
        for(int i=0;i<genres.length;i++){
            genre.putIfAbsent(genres[i], new ArrayList<>());
            genre.get(genres[i]).add(new int[]{plays[i], i});
            cnt.put(genres[i], cnt.getOrDefault(genres[i],0) + plays[i]);
        }
        
        List<String> tmp = new ArrayList<>(cnt.keySet());
        tmp.sort((a,b)-> cnt.get(b)-cnt.get(a));
        
        List<Integer> ans = new ArrayList<>();
        for(String g: tmp){
            List<int[]> cur = genre.get(g);
            
            cur.sort((a,b)-> {
                if(b[0]!=a[0]) return b[0]-a[0];
                return a[1]-b[1];
            });
            
            for(int i=0;i<Math.min(2,cur.size());i++){
                ans.add(cur.get(i)[1]);
            }
        }
        
        return ans;
    }
}